### PR TITLE
Update foreman inventory to use foreman's inventory report

### PR DESCRIPTION
**Pull request by sjha4**
##### SUMMARY
Currently the foreman inventory script relies on API calls to foreman to fetch hosts, details and facts. This is achieved my making one call to get the list of hosts, followed by 2 calls per host to fetch details and facts. This is slow and sometimes take hours to complete.

Proposed change is to implement a report which provides all the required data from foreman (WIP here: https://github.com/theforeman/foreman_ansible/pull/299) Following this, the inventory scripts needs to API calls. First to generate the report and the second to fetch the generated data.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
Foreman Inventory

##### ADDITIONAL INFORMATION
Related awx PR: https://github.com/ansible/awx/pull/4758

```paste below

```
Work in Progress:
Before Change: https://codebeautify.org/jsonviewer/cb1e7f3e
After Change: https://codebeautify.org/jsonviewer/cbbae5d2

### CODE REVIEW

**Comment by sjha4** [comment]
Based on discussions here: https://github.com/ansible/ansible/pull/62542 , this change should probably live in plugins/inventory.

**Comment by AlanCoding** [comment]
No, I think you want this PR merged if you want the content to make it into AWX _soon_.

You might want to _also_ consider putting the content in the inventory plugin.

File: contrib/inventory/foreman.py
Comment (ares) [inline contrib/inventory/foreman.py]: this should use the new API without hardcoded id now

File: contrib/inventory/foreman.py
Comment (ekohl) [inline contrib/inventory/foreman.py:269]: Interesting approach. I was thinking about catching a 404 on the `schedule_report`.

File: contrib/inventory/foreman.py
Comment (ekohl) [inline contrib/inventory/foreman.py:299]: It'd also be good to verify the response code is 201 (which I assume it actually returns). If the plugin is not installed, it will still return a 404.

File: contrib/inventory/foreman.py
Comment (ekohl) [inline contrib/inventory/foreman.py]:
```suggestion
    def _post_request(self):
```

File: contrib/inventory/foreman.py
Comment (ekohl) [inline contrib/inventory/foreman.py]: I'd use a dict comprehension if this can contain more than 2 elements:
```suggestion
            content_facet_attributes = {cf[0]: cf[1] for cf in content_facet_data}
```

If it's just two attributes it's easier to use the dict constructor:
```suggestion
            content_facet_attributes = dict(host.get('content_facet_attributes', {}))
```

However, the data types here are quite weird. The above line implies it's always a dict but that means the key is a tuple and the value is disregarded. It'd be good to add a test for this.

File: contrib/inventory/foreman.py
Comment (ekohl) [inline contrib/inventory/foreman.py]: Here I have the same comments as above.

File: contrib/inventory/foreman.py
Comment (ekohl) [inline contrib/inventory/foreman.py]: I find the inconsistent capitalization very confusing. There's `Name` and `Facts` but also `all_parameters`. Here we have `environment` but also `Location` and `Organization`. Ideally the original API response would be normalized.

File: contrib/inventory/foreman.py
Comment (ekohl) [inline contrib/inventory/foreman.py:448]:
```suggestion
            groupby = {k: self.to_safe(to_text(v)) for k, v in params.items()}
```

File: contrib/inventory/foreman.py
Comment (ekohl) [inline contrib/inventory/foreman.py]:
```suggestion

```

File: contrib/inventory/foreman.py
Comment (ekohl) [inline contrib/inventory/foreman.py]: I'd also include a check for server errors (500)

File: contrib/inventory/foreman.py
Comment (ekohl) [inline contrib/inventory/foreman.py]: Probably cleaner to use a different variable than before to make sure you're not using the incorrect response.

File: contrib/inventory/foreman.py
Comment (ekohl) [inline contrib/inventory/foreman.py]: If the result was a 4xx or 5xx then I think this fails.

File: contrib/inventory/foreman.py
Comment (ekohl) [inline contrib/inventory/foreman.py]: Just to be sure add a `if not host_data: ...`

**Review by ekohl — CHANGES_REQUESTED** [review]

**Comment by sivel** [comment]
We are no longer accepting changes to the contrib inventory scripts. Contrib inventory scripts are deprecated and replaced by inventory plugins. The foreman inventory plugin can be found at https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/inventory/foreman.py

**Comment by ekohl** [comment]
@sivel I agree, but can ansibot actually do this automatically?

**Comment by sjha4** [comment]
@sivel : Question around deprecating this. How does AWX handle that? Would the changes in the inventory plugin flow to AWX?

**Comment by sivel** [comment]
I cannot speak to the development practices of AWX, but inventory plugins should be usable from within AWX. But I'd have to defer to the AWX team to further answer your questions.

**Comment by AlanCoding** [comment]
One question that should be asked - is this what https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/inventory/foreman.py does? That is officially maintained, and we would like to keep logic closely aligned to that because we are going to switch to it at some point.

**Comment by ws4e** [comment]
Guys this needs to be better handled.  The foreman.py inventory plugin for AWX and command line ansible are quite different, with the AWX one it seems being further along and more maintained (eg has the ansible_ssh_host feature I use myself), yet now the AWX guys are saying any updates need to happen here.    I think these should be first synchronized(the awx version should replace this one most likely) before moving forward.

**Comment by ares** [comment]
I agree we need to unify these two. Note that the work done here completely changes the way we generate inventory and should increase performance. I understand the plugins/inventory/foreman.py is the new place, but we also want to get it to earliest AWX possible. And my understaind is that contrib/inventory/foreman.py is the inventory, that's currently being synchronized and used in AWX. Can we finish the task here and then continue with unification, leading to plugins/inventory/foreman.py be the only one?

File: contrib/inventory/foreman.py
Comment (sjha4) [reply contrib/inventory/foreman.py:448]: I tried doing this and even though this works for me with python 3.6 , the CI isn't happy with this on python 2.6 and throws an error.

**Comment by kdelee** [comment]
I tried this out and I am getting this error:
```
raise Exception("Error scheduling inventory report on foreman. Please check
foreman logs!") Exception: Error scheduling inventory report on foreman. Please
check foreman logs!
```

**Comment by sjha4** [comment]
> I tried this out and I am getting this error:
> 
> ```
> raise Exception("Error scheduling inventory report on foreman. Please check
> foreman logs!") Exception: Error scheduling inventory report on foreman. Please
> check foreman logs!
> ```

Hey @kdelee : Could you check the /api/v2/status on your foreman instance?

**Comment by ares** [comment]
I've seen Alan experienced the same and when I was checking the foreman instance logs, I've seen authentication issue. Perhaps wrong credentials?

**Comment by sjha4** [comment]
@ares: Related to this, we have this assumption that foreman users on 1.24 will necessarily have foreman-ansible and the inventory report available. Right now the check is only foreman version 1.24. You think we should also fall back to the old api method if the report scheduling fails? Or should we just let users configure that with the use_reports_api param?

**Comment by ares** [comment]
The requirements are:
foreman 1.24
katello 3.9.0
foreman_ansible 3.1.0

but you can simply check for the API availability and foreman version being 1.24+

File: contrib/inventory/foreman.ini
Comment (ekohl) [inline contrib/inventory/foreman.ini]: Why the inconsistency in casing?

File: contrib/inventory/foreman.ini
Comment (ekohl) [inline contrib/inventory/foreman.ini]: Personally I consider IPv4 legacy and IPv6 default. Can we enable this by default?

File: contrib/inventory/foreman.py
Comment (ekohl) [inline contrib/inventory/foreman.py]: The braces aren't needed
```suggestion
        if not self.foreman_use_reports_api:
```

File: contrib/inventory/foreman.py
Comment (ekohl) [inline contrib/inventory/foreman.py]: The braces are not needed
```suggestion
        if not response:
```

However, I don't think `session.get` ever returns `None`. You should also check if it's not an error (HTTP 500) because otherwise `response.json()` later will fail.

**Comment by sjha4** [comment]
So I have updated both the ansible and the awx PR to do this: For foreman >= 1.24 where we want to use the reports API, If there are any errors in scheduling and downloading report, fallback to the old API..

File: contrib/inventory/foreman.py
Comment (ekohl) [reply contrib/inventory/foreman.py:448]: Oh yes, 2.6 didn't have dict comprehensions. Do we still need to care about that? setup.py indicates only 2.7 should be supported.

File: contrib/inventory/foreman.ini
Comment (sjha4) [reply contrib/inventory/foreman.ini]: These correspond to the report parameters that we have on the foreman report. But it's possible to change it here if you feel strongly about it.

File: contrib/inventory/foreman.py
Comment (sjha4) [reply contrib/inventory/foreman.py]: The request.response evaluates to True if the status code is between 200 and 400, and False otherwise. So this will be able to handle 500s..

File: contrib/inventory/foreman.ini
Comment (ekohl) [reply contrib/inventory/foreman.ini]: it does feel very inconsistent as a user. I didn't like the capitalization in the report itself either.

File: contrib/inventory/foreman.py
Comment (ekohl) [reply contrib/inventory/foreman.py]: Ah, I didn't know that. That actually makes sense.

File: contrib/inventory/foreman.ini
Comment (ares) [reply contrib/inventory/foreman.ini]: It's not fully supported in Foreman, so it make sense to default to ipv4 to me. Also these values reflect the default in the inventory template on Foreman side, I'd start with the change there.

File: contrib/inventory/foreman.py
Comment (ekohl) [inline contrib/inventory/foreman.py:279]: TIL you can index a tuple with a boolean:
```
In [1]: ("no", "yes")[True]                                                                                                                                                           
Out[1]: 'yes'
```

I do wonder if we can get the report to accept real booleans since they're JSON encoded.

File: contrib/inventory/foreman.ini
Comment (ekohl) [reply contrib/inventory/foreman.ini]: I agree it should be changed in the template as well. To be clear: I'd expect dual stack support by default so both IPv4 and IPv6 should default to true IMHO.

File: contrib/inventory/foreman.py
Comment (sjha4) [reply contrib/inventory/foreman.py:448]: The CI has tests around versions which fail in 2.6..

File: contrib/inventory/foreman.ini
Comment (sjha4) [reply contrib/inventory/foreman.ini]: Updated these to be lower case..

**Comment by sjha4** [comment]
@ekohl @ares : Do we need any changes here related to the 2 pending review comments? 
a) Default values for ipv4 and ipv6 
b) Passing booleans to report.

**Comment by ares** [comment]
I'd be happy to get this in soon and iterate on it, once we unify it woth tower version

**Comment by AlanCoding** [comment]
What's the status of this? As I put in the review of the related AWX PR I am happy with the content and would like to get it moving. Having this merged first would be easiest.

**Comment by sjha4** [comment]
@ares , @AlanCoding : we do need acks on this PR if this looks ok to you?

**Comment by ares** [comment]
Ack from my side, but I don't have merge permissions, all I can say is shipit :-)

**Comment by AlanCoding** [comment]
I have non-blocking comments at https://github.com/ansible/awx/pull/4988#issuecomment-555787856, but I defer to your judgement on those points. 👍 from me.

**Comment by sjha4** [comment]
@AlanCoding : I will update the Tower PR with the diff with the version that gets merged. The changes are minor and mostly linting related :+1: 
@Ansible Team: Does this look ok to merge?

File: contrib/inventory/foreman.ini
Comment (ekohl) [inline contrib/inventory/foreman.ini]: Nit: this relies on the `foreman_ansible` plugin. Users who have 1.24 but not the plugin will see an error and be confused. Plesae add it as a note. Also add a space after `#` and wrap the line for readability.

File: contrib/inventory/foreman.ini
Comment (ekohl) [inline contrib/inventory/foreman.ini]:
```suggestion
# want_content_facet_attributes = False
```

File: contrib/inventory/foreman.ini
Comment (sjha4) [reply contrib/inventory/foreman.ini]: Updated.

**Comment by sjha4** [comment]
@ekohl : Updated as per your suggestions.
@AlanCoding : Updated AWX PR with the version here.

**Review by jillr — APPROVED** [review]
LGTM

**Comment by ares** [comment]
Thank you @sjha4 and everyone who was involved!

**Pull request merged**
